res/controllers/Numark-Mixtrack-Pro-FX-scripts.js

```diff
@@ -0,0 +1,746 @@
+MixtrackProFX.autoLoopSizes = [
+    "0.0625",
+    "0.125",
+    "0.25",
+    "0.5",
+    "1",
+    "2",
+    "4",
+    "8"
+];
+
+// beatjump values, for available values see:
+// https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#control-[ChannelN]-beatjump_X_forward
+// underscores (_) at the end are needed because numeric values (e.g. 8) have two underscores (e.g. beatjump_8_forward),
+// but "beatjump_forward"/"beatjump_backward" have only one underscore
+MixtrackProFX.beatJumpValues = [
+    "0.0625_",
+    "0.125_",
+    "0.25_",
+    "0.5_",
+    "1_",
+    "2_",
+    "", // "beatjump_forward"/"beatjump_backward" - jump by the value selected in Mixxx GUI (4 by default)
+    "8_"
+];
+
+// pad modes, don't touch
+MixtrackProFX.PadModeControls = {
+    HOTCUE: 0x00,
+    AUTOLOOP: 0x0D,
+    FADERCUTS: 0x07,
+    SAMPLE1: 0x0B,
+    BEATJUMP: 0x02,
+    SAMPLE2: 0x0F
+};
+
+// state variables, don't touch
+MixtrackProFX.shifted = false;
+MixtrackProFX.scratchModeEnabled = [true, true];
+
+// a helper function to convert value to binary scale (0 or 1)
+MixtrackProFX.scaleToBinaryInput = function(value) {
+    print("debug MixtrackProFX.scaleToBinaryInput");
+    if (value > 0) {
+        return 1;
+    }
+    return 0;
+};
+
+MixtrackProFX.init = function() {
+    // initialize component containers
+    MixtrackProFX.deck = new components.ComponentContainer();
+    MixtrackProFX.effect = new components.ComponentContainer();
+    var i;
+    for (i = 0; i < 2; i++) {
+        MixtrackProFX.deck[i] = new MixtrackProFX.Deck(i + 1);
+        MixtrackProFX.effect[i] = new MixtrackProFX.EffectUnit(i + 1);
+    }
+
+    MixtrackProFX.browse = new MixtrackProFX.Browse();
+    MixtrackProFX.gains = new MixtrackProFX.Gains();
+
+    // send sysexes
+    var exitDemoSysex = [0xF0, 0x7E, 0x00, 0x06, 0x01, 0xF7];
+    midi.sendSysexMsg(exitDemoSysex, exitDemoSysex.length);
+
+    var statusSysex = [0xF0, 0x00, 0x20, 0x7F, 0x03, 0x01, 0xF7];
+    midi.sendSysexMsg(statusSysex, statusSysex.length);
+
+    // enables 4 bottom pads "fader cuts"
+    var faderCutSysex = [0xF0, 0x00, 0x20, 0x7F, 0x03, 0xF7];
+    midi.sendSysexMsg(faderCutSysex, faderCutSysex.length);
+
+    // initialize leds
+    for (i = 0; i < 2; i++) {
+        midi.sendShortMsg(0x90 + i, 0x07, 0x7F); // scratch
+        midi.sendShortMsg(0x94 + i, 0x00, 0x7F); // hotcue
+        midi.sendShortMsg(0x94 + i, 0x0D, 0x01); // auto loop
+        midi.sendShortMsg(0x94 + i, 0x07, 0x01); // "fader cuts"
+        midi.sendShortMsg(0x94 + i, 0x0B, 0x01); // sample1
+
+        // shifted leds
+        midi.sendShortMsg(0x94 + i, 0x0F, 0x01); // sample2
+        midi.sendShortMsg(0x94 + i, 0x02, 0x01); // beatjump
+    }
+
+    engine.makeConnection("[Channel1]", "VuMeter", MixtrackProFX.vuCallback);
+    engine.makeConnection("[Channel2]", "VuMeter", MixtrackProFX.vuCallback);
+
+    MixtrackProFX.deck.reconnectComponents();
+    MixtrackProFX.effect.reconnectComponents();
```

Comment:
I added it only to set initial state of button lights (to 0x01). Is it better to add lots of calls to midi.sendShortMsg(...) instead of reconnecting components?

Reply:
Ah, well reconnectComponents disconnects, then reconnects, and then triggers all bound CO callbacks. You are essentially only interested in the trigger part. Have you tried to just remove the reconnectComponents call here entirely though? Because technically all components already trigger upon construction. See this snippet from the Component constructor:
```javascript
if (this.outConnect !== false && this.group !== undefined && this.outKey !== undefined) {
    this.connect();
    if (this.outTrigger !== false) {
        this.trigger();
    }
}
```
Thats why I'm questioning whether manually triggering each component is really necessary.

Reply:
That is theoretically more ideal. Though the fact that it seems to be necessary seems to be a deficiency in componentJS. I'll investigate in the future.

Reply:
So... should I go with .trigger()?

Reply:
Yes, its more verbose in code but the better solution from a technical viewpoint.